infer/default_provider.go

```diff
@@ -0,0 +1,251 @@
+	components    []InferredComponent
+	functions     []InferredFunction
+	config        InferredConfig
+	moduleMap     map[tokens.ModuleName]tokens.ModuleName
+}
+
+// NewDefaultProvider creates an inferred provider which fills as many defaults as possible.
+//
+// A base set of defaults are provided to create a minimal provider, but can be initially overridden by
+// passing in an [Options] struct. Further customization can be done by chaining method calls
+// on the returned [DefaultProvider] object.
+//
+// This is an example of how to create a simple provider with a single component resource:
+//
+//	type RandomComponent struct {
+//		pulumi.ResourceState
+//		RandomComponentArgs
+//	 	Password        pulumi.StringOutput `pulumi:"password"`
+//	}
+//
+//	type RandomComponentArgs struct {
+//		Length pulumi.IntInput `pulumi:"length"`
+//	}
+//
+//	func NewMyComponent(ctx *pulumi.Context, name string,
+//			compArgs RandomComponentArgs, opts ...pulumi.ResourceOption) (*RandomComponent, error) {
+//		// Define your component constructor logic here.
+//	}
+//
+//	func main() {
+//		err := infer.NewDefaultProvider(nil).
+//			WithName("go-components").
+//			WithVersion("v0.0.1").
+//			WithComponents(
+//				infer.Component(NewMyComponent),
+//			).
+//			BuildAndRun()
+//	}
+func NewDefaultProvider(opts *Options) *DefaultProvider {
+	defaultMetadata := schema.Metadata{
+		LanguageMap: map[string]any{
+			"nodejs": map[string]any{
+				"respectSchemaVersion": true,
+			},
+			"go": map[string]any{
+				"generateResourceContainerTypes": true,
+				"respectSchemaVersion":           true,
+			},
+			"python": map[string]any{
+				"requires": map[string]any{
+					"pulumi": ">=3.0.0,<4.0.0",
+				},
+				"respectSchemaVersion": true,
+			},
+			"csharp": map[string]any{
+				"packageReferences": map[string]any{
+					"Pulumi": "3.*",
+				},
+				"respectSchemaVersion": true,
+			},
+		},
+	}
+
+	if opts == nil {
+		return &DefaultProvider{
+			metadata: defaultMetadata,
+		}
+	}
+
+	dp := &DefaultProvider{
+		metadata: schema.Metadata{
+			LanguageMap:       opts.Metadata.LanguageMap,
+			Description:       opts.Metadata.Description,
+			DisplayName:       opts.Metadata.DisplayName,
+			Keywords:          opts.Metadata.Keywords,
+			Homepage:          opts.Metadata.Homepage,
+			Repository:        opts.Metadata.Repository,
+			Publisher:         opts.Metadata.Publisher,
+			LogoURL:           opts.Metadata.LogoURL,
+			License:           opts.Metadata.License,
+			PluginDownloadURL: opts.Metadata.PluginDownloadURL,
+		},
+		resources:  opts.Resources,
+		components: opts.Components,
+		functions:  opts.Functions,
+		config:     opts.Config,
+		moduleMap:  opts.ModuleMap,
+	}
+
+	if dp.metadata.LanguageMap == nil {
+		dp.metadata.LanguageMap = defaultMetadata.LanguageMap
+	}
+
+	return dp
+}
+
+// WithResources adds the given custom resources to the provider.
+func (dp *DefaultProvider) WithResources(resources ...InferredResource) *DefaultProvider {
```

Comment:
Other's might argue, but I'd prefer that this didn't mutate.
```suggestion
func (dp DefaultProvider) WithResources(resources ...InferredResource) *DefaultProvider {
```

Reply:
Using receivers make method chaining easier for users, and is more Go idiomatic from what I've seen in other repositories. I'm more favorable towards using receivers here.